File: monsoon3.0/src/main/java/wtf/monsoon/impl/module/movement/flight/VerusFlight.java
package wtf.monsoon.impl.module.movement.flight;

import io.github.nevalackin.homoBus.Listener;
import io.github.nevalackin.homoBus.annotations.EventLink;
import net.minecraft.block.BlockAir;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.network.play.client.C08PacketPlayerBlockPlacement;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.BlockPos;
import wtf.monsoon.Wrapper;
import wtf.monsoon.api.module.Module;
import wtf.monsoon.api.setting.ModeProcessor;
import wtf.monsoon.api.setting.Setting;
import wtf.monsoon.api.util.misc.Timer;
import wtf.monsoon.impl.event.EventBlockCollide;
import wtf.monsoon.impl.event.EventMove;
import wtf.monsoon.impl.event.EventPreMotion;
import wtf.monsoon.impl.module.combat.Aura;
import wtf.monsoon.impl.module.movement.Flight;

public class VerusFlight extends ModeProcessor {

    private final Setting<VerusMode> verusMode = new Setting<>("Verus Mode", VerusMode.AIRWALK)
            .describedBy("The mode of the Verus flight.");

    private final Setting<Boolean> fastVerus = new Setting<>("Fast", false)
            .describedBy("Whether to go faster on Verus.")
            .visibleWhen(() -> verusMode.getValue() == VerusMode.BRUE);

    private int ticks;

    private Timer timer = new Timer();

    public VerusFlight(Module parentModule) {
        super(parentModule);
    }

    @Override
    public void onEnable() {
        super.onEnable();
        if (verusMode.getValue() == VerusMode.BRUE) {
            ticks = 0;
            player.setSpeed(0);
            mc.thePlayer.setVelocity(0, 0, 0);
        }
    }

    @Override
    public void onDisable() {
        super.onDisable();
        mc.thePlayer.setVelocity(0, 0, 0);
        player.setSpeed(0);
    }

    @EventLink
    public final Listener<EventPreMotion> eventPreMotionListener = e -> {
        switch (verusMode.getValue()) {
            case AIRWALK:
                player.setOnGround(true);
                e.setOnGround(mc.thePlayer.ticksExisted % 2 == 0);
                mc.thePlayer.motionY = 0;
                e.setY(Math.round(mc.thePlayer.posY));
                mc.thePlayer.sendQueue.addToSendQueue(new C08PacketPlayerBlockPlacement(new BlockPos(mc.thePlayer.prevPosX, mc.thePlayer.posY - 1, mc.thePlayer.prevPosZ), 1, new ItemStack(Blocks.stone), 1, 1, 1));
                break;
            case BRUE:
                break;
        }
    };

    @EventLink
    public final Listener<EventMove> eventMoveListener = e -> {
            switch (verusMode.getValue()) {
                case BRUE:
                    if (ticks % 14 == 0 && player.isOnGround()) {
                        player.setSpeed(e, 1);
                        e.setY(0.42F);
                        mc.thePlayer.motionY = -(mc.thePlayer.posY - (mc.thePlayer.posY - (mc.thePlayer.posY % 0.015625)));
                    } else {
                        if (fastVerus.getValue()) {
                            if (player.isOnGround()) {
                                player.setSpeed(e, player.getBaseMoveSpeed() * 2.65f);
                            } else player.setSpeed(e, 0.41);
                        } else {
                            if (player.isOnGround()) {
                                if (mc.thePlayer.moveStrafing == 0) {
                                    float multiplier = (float) (1.0f + (mc.thePlayer.motionY < 0f ? mc.thePlayer.motionY * -6f : mc.thePlayer.motionY * 6f));
                                    if (mc.thePlayer.hurtTime > 0)
                                        player.setSpeed(e, player.getBaseMoveSpeed() * 1.7f);
                                    else player.setSpeed(e, player.getBaseMoveSpeed() * multiplier);
                                } else {
                                    player.setSpeed(e, player.getSpeed() * 0.9f);
                                }
                            } else player.setSpeed(e, 0.41);
                        }
                    }
                    ticks++;
                    break;
                case AIRWALK:
                    break;
            }
    };

    @EventLink
    public final Listener<EventBlockCollide> eventBlockCollideListener = e -> {
        switch (verusMode.getValue()) {
            case BRUE:
                if (e.getBlock() instanceof BlockAir && !mc.thePlayer.isSneaking()) {
                    final double x = e.getX(), y = e.getY(), z = e.getZ();

                    if (y < mc.thePlayer.posY) {
                        e.setCollisionBoundingBox(AxisAlignedBB.fromBounds(-15, -1, -15, 15, 1, 15).offset(x, y, z));
                    }
                }
                break;

            case AIRWALK:
                break;
        }
    };

    @Override
    public Setting[] getModeSettings() {
        return new Setting[] { verusMode, fastVerus };
    }

    private enum VerusMode {
        BRUE,
        AIRWALK
    }

}

File: monsoon-3.x-v2/src/main/java/wtf/monsoon/client/modules/movement/flight/VerusFlight.kt
package wtf.monsoon.client.modules.movement.flight

import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.block.BlockAir
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.network.play.client.C08PacketPlayerBlockPlacement
import net.minecraft.util.AxisAlignedBB
import net.minecraft.util.BlockPos
import wtf.monsoon.backend.module.MulticlassModule
import wtf.monsoon.backend.module.mode.ModuleMode
import wtf.monsoon.backend.setting.Setting
import wtf.monsoon.client.event.EventBlockCollide
import wtf.monsoon.client.event.EventMove
import wtf.monsoon.client.event.EventPacket
import wtf.monsoon.client.event.EventPreMotion
import wtf.monsoon.client.modules.movement.Flight
import wtf.monsoon.client.util.misc.Stopwatch
import kotlin.math.roundToInt

class VerusFlight(name: String, parent: MulticlassModule) : ModuleMode<Flight>(name, parent) {

    private val verusMode: Setting<Mode> = Setting("Verus Mode", Mode.AIRWALK)

    private val fastVerus: Setting<Boolean> = Setting("Fast", false)
        .visibleWhen { verusMode.getValue() === Mode.BRUE }

    private var ticks = 0

    private val timer: Stopwatch = Stopwatch()

    init {
       this.registerSettings(verusMode, fastVerus)
    }


    override fun enable() {
        super.enable()
        if (verusMode.getValue() === Mode.BRUE) {
            ticks = 0
            player.setSpeed(0.0)
            mc.thePlayer.setVelocity(0.0, 0.0, 0.0)
        }
    }

    override fun disable() {
        super.disable()
        mc.thePlayer.setVelocity(0.0, 0.0, 0.0)
        player.setSpeed(0.0)
    }

    @EventListener
    val preMotion = fun(e: EventPreMotion) {
        when (verusMode.getValue()) {
            Mode.AIRWALK -> {
                mc.thePlayer.onGround = true
                e.onGround = mc.thePlayer.ticksExisted % 2 == 0
                mc.thePlayer.motionY = 0.0
                e.y = mc.thePlayer.posY.roundToInt().toDouble()
                mc.thePlayer.sendQueue.addToSendQueue(
                    C08PacketPlayerBlockPlacement(
                        BlockPos(
                            mc.thePlayer.prevPosX,
                            mc.thePlayer.posY - 1,
                            mc.thePlayer.prevPosZ
                        ), 1, ItemStack(Blocks.stone), 1f, 1f, 1f
                    )
                )
            }

            Mode.BRUE -> {}
        }
    }

    @EventListener
    val eventMove = fun(e: EventMove) {
        when (verusMode.getValue()) {
            Mode.BRUE -> {
                if (ticks % 14 == 0 && player.isOnGround) {
                    player.setSpeed(e, 1f)
                    e.y = 0.42
                    mc.thePlayer.motionY = -(mc.thePlayer.posY - (mc.thePlayer.posY - mc.thePlayer.posY % 0.015625))
                } else {
                    if (fastVerus.getValue()) {
                        if (player.isOnGround) {
                            player.setSpeed(e, player.baseMoveSpeed * 2.65f)
                        } else player.setSpeed(e, 0.41)
                    } else {
                        if (player.isOnGround) {
                            if (mc.thePlayer.moveStrafing == 0f) {
                                val multiplier =
                                    (1.0f + if (mc.thePlayer.motionY < 0f) mc.thePlayer.motionY * -6f else mc.thePlayer.motionY * 6f).toFloat()
                                if (mc.thePlayer.hurtTime > 0) player.setSpeed(
                                    e,
                                    player.baseMoveSpeed * 1.7f
                                ) else player.setSpeed(e, player.baseMoveSpeed * multiplier)
                            } else {
                                player.setSpeed(e, player.speed * 0.9f)
                            }
                        } else player.setSpeed(e, 0.41)
                    }
                }
                ticks++
            }

            Mode.AIRWALK -> {}
        }
    }

    @EventListener
    val eventBlockCollide = fun(e: EventBlockCollide) {
        when (verusMode.getValue()) {
            Mode.BRUE ->
                if (e.block is BlockAir && !mc.thePlayer.isSneaking) {
                    val x: Double = e.x!!
                    val y: Double = e.y!!
                    val z: Double = e.z!!
                    if (y < mc.thePlayer.posY)
                        e.collisionBoundingBox = AxisAlignedBB.fromBounds(-15.0, -1.0, -15.0, 15.0, 1.0, 15.0).offset(x, y, z)
                }

            Mode.AIRWALK -> {}
        }
    }

    private enum class Mode {
        BRUE, AIRWALK
    }

}